### PR TITLE
support for .icns files

### PR DESCRIPTION
**Pull request by typemytype**

### CODE REVIEW

**Comment by typemytype** [comment]
still todo: write test

File: drawBot/context/icnsContext.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+class ICNSContext(ImageContext):
+
+    fileExtensions = ["icns"]
+
+    allowedPageSizes = [16, 32, 128, 256, 512, 1024]
+
+    def _writeDataToFile(self, data, path, options):
+        # create a iconset folder
+        iconsetPath = tempfile.mkdtemp(suffix=".iconset")
```
Comment (justvanrossum) [inline drawBot/context/icnsContext.py]: The rest of the code should be in a try/finally, nw there's a chance this temp folder won't get deleted. Better yet, use `tempfile.TemporaryDirectory()` as a context manager, if Py3-only is ok.

File: drawBot/context/icnsContext.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        # get the complete pdf
+        pdfDocument = Quartz.PDFDocument.alloc().initWithData_(data)
+        pageCount = pdfDocument.pageCount()
+        # set the image resolution
+        options["imageResolution"] = 72
+        # make a copy and alter the resolution
+        options_2x = dict(options)
+        options_2x["imageResolution"] = 144
+        # start loop over all pages
+        for index in range(0, pageCount):
```
Comment (justvanrossum) [inline drawBot/context/icnsContext.py]: `range(pageCount)` will do the same.

File: drawBot/context/icnsContext.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            # get the pdf page
+            page = pdfDocument.pageAtIndex_(index)
+            # get the pdf page, this acts also as pdf document...
+            pageData = page.dataRepresentation()
+            # extract the size of the page
+            _, (w, h) = page.boundsForBox_(Quartz.kPDFDisplayBoxArtBox)
+            w = int(round(w))
+            h = int(round(h))
+            # dont allow any other size, the command iconutil will not work otherwise
+            if w not in self.allowedPageSizes or h not in self.allowedPageSizes:
```
[the remark this answers is not in the document]
Comment (justvanrossum) [reply drawBot/context/icnsContext.py]: Or: `if w not in self.allowedPageSizes or w != h:`

File: drawBot/context/icnsContext.py (hunk 1 above)
Comment (typemytype) [reply drawBot/context/icnsContext.py]: we have to setup a plan how long py2 is been supported, it is for now...

**Pull request merged**